compiler/rustc_feature/src/lib.rs

```diff
@@ -74,11 +85,20 @@ impl UnstableFeatures {
         // Returns whether `krate` should be counted as unstable
         let is_unstable_crate =
             |var: &str| krate.is_some_and(|name| var.split(',').any(|new_krate| new_krate == name));
-        // `true` if we should enable unstable features for bootstrapping.
-        let bootstrap =
-            std::env::var("RUSTC_BOOTSTRAP").is_ok_and(|var| var == "1" || is_unstable_crate(&var));
-        match (disable_unstable_features, bootstrap) {
-            (_, true) => UnstableFeatures::Cheat,
+
+        let bootstrap_kind = match std::env::var("RUSTC_BOOTSTRAP") {
+            Ok(ref val) if val == "1" || is_unstable_crate(&val) => RustcBootstrapKind::Cheat,
+            // `RUSTC_BOOTSTRAP=-1` instead of `0` to make it clear that we are *forcing* us to
+            // pretend to be a stable compiler with no features enabled.
+            Ok(ref val) if val == "-1" => RustcBootstrapKind::ForceStable,
+            _ => RustcBootstrapKind::Unknown,
+        };
+
+        match (disable_unstable_features, bootstrap_kind) {
+            (_, RustcBootstrapKind::Cheat) => UnstableFeatures::Cheat,
+            // Pretend very hard that we are a stable compiler, don't allow any unstable features
+            // through.
+            (_, RustcBootstrapKind::ForceStable) => UnstableFeatures::Disallow,
```

Comment:
I think the existence of `bootstrap_kind` and `RUSTC_BOOTSTRAP_KIND` adds confusion, because it's an entire enum that only exists very briefly as an implementation detail.

Given how the logic works, I think it might be better for the `RUSTC_BOOTSTRAP=1` and `RUSTC_BOOTSTRAP=-1` cases to just immediately return `UnstableFeatures::Cheat` or `UnstableFeatures::Disallow` respectively.

Then the rest of the function would implement the “normal” logic for distinguishing stable/nightly compilers, when `RUSTC_BOOTSTRAP` is not set to a relevant value.

Reply:
Thanks, I just collapsed this into a simpler match. lmw if it looks simpler to you.

Reply:
Updated to early return on `RUSTC_BOOTSTRAP=1`, `RUSTC_BOOTSTRAP=crate_list` and `RUSTC_BOOTSTRAP=-1` when `RUSTC_BOOTSTRAP` is set.